WIP: AVX2 Float 32x32 SP

deeptiag1 opened the pull request:
Submitting to get feedback on the blocking sizes. Thanks!

alheinecke commented on src/generator_gemm_sse_avx_avx2_avx512.c:
```diff
@@ -634,6 +636,43 @@ LIBXSMM_API_INTERN unsigned int libxsmm_generator_gemm_sse_avx_avx2_avx512_get_i
+        l_use_masking_a_c = 1;
+      }
+    }
+    // if ( i_xgemm_desc->m >= 64 ) {
```
Please use C comments and not C++ comments

deeptiag1 replied:
Removed comments

alheinecke commented on src/generator_common.c:
```diff
@@ -1166,10 +1166,10 @@ LIBXSMM_API_INTERN unsigned int libxsmm_compute_equalized_blocking(
-  *o_block_1 = l_n1;
-  *o_range_2 = l_N2;
-  *o_block_2 = l_n2;
+  *o_range_1 = l_N1; //78
```
not sure, why these comments are there, please use C comments instead of C++ if these comments are temp. needed

deeptiag1 replied:
Removed comments.

alheinecke commented on src/generator_gemm_sse_avx_avx2_avx512.c:
```diff
@@ -634,6 +636,43 @@ LIBXSMM_API_INTERN unsigned int libxsmm_generator_gemm_sse_avx_avx2_avx512_get_i
+    } else {
+      l_m_blocking = i_xgemm_desc->m;
+      /* in case we don't have a full vector length, we use masking */
+      if ( l_m_blocking % 8 != 0 ) {
```
that should be mod 4

sbhatta3 replied:
fixed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -753,7 +753,16 @@ void libxsmm_x86_instruction_vec_mask_move( libxsmm_generated_code* io_generated
   }
 }
-
+/*
```
not sure why this code was added as comment

deeptiag1 replied:
Removed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -909,11 +918,15 @@ void libxsmm_x86_instruction_vec_move( libxsmm_generated_code* io_generated_code
 
       /* determine encoder */
+      //Change by D-
       if ( io_generated_code->arch < LIBXSMM_X86_AVX512) {
```
I would change this one to < LIBXSMM_X86_AVX512_VL256

deeptiag1 replied:
Fixed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -909,11 +918,15 @@ void libxsmm_x86_instruction_vec_move( libxsmm_generated_code* io_generated_code
       } else if ( io_generated_code->arch < LIBXSMM_X86_AVX ) {
         l_encoder_arch = 0;
       }
+      if ( io_generated_code->arch == LIBXSMM_X86_AVX512_VL256) {
```
Then this one can be removed

deeptiag1 replied:
Removed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -1557,6 +1570,9 @@ void libxsmm_x86_instruction_vec_compute_3reg_mask_sae_imm8( libxsmm_generated_c
     } else if ( io_generated_code->arch < LIBXSMM_X86_AVX ) {
       l_encoder_arch = 0;
     }
+    if ( io_generated_code->arch == LIBXSMM_X86_AVX512_VL256) {
```
see above

deeptiag1 replied:
Changed as per feedback

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -2040,6 +2056,9 @@ void libxsmm_x86_instruction_vec_compute_mem_2reg_mask_imm8( libxsmm_generated_c
     } else if ( io_generated_code->arch < LIBXSMM_X86_AVX ) {
       l_encoder_arch = 0;
     }
+    if ( io_generated_code->arch == LIBXSMM_X86_AVX512_VL256) {
```
see above

deeptiag1 replied:
Changed as recommended

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -3023,7 +3042,7 @@ void libxsmm_x86_instruction_vex_evex_mask_mov( libxsmm_generated_code* io_gener
                                                 const unsigned int      i_mask_reg_number,
                                                 const unsigned int      i_is_store ) {
-  if ( io_generated_code->arch >= LIBXSMM_X86_AVX512 ) {
+  if (( io_generated_code->arch >= LIBXSMM_X86_AVX512 ) || ( io_generated_code->arch >= LIBXSMM_X86_AVX512_VL256 )) {
```
see above >= LIBXSMM_X86_AVX512_VL256 incl.  >= LIBXSMM_X86_AVX512

deeptiag1 replied:
Changed as recommended

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -762,12 +771,12 @@ void libxsmm_x86_instruction_vec_move( libxsmm_generated_code* io_generated_code
-                                       const unsigned int      i_mask_reg_number,
-                                       const unsigned int      i_use_zero_masking,
-                                       const unsigned int      i_is_store )
+                                       const int               i_displacement,// ((l_n * i_xgemm_desc->ldc) + (l_m * (i_micro_kernel_config->vector_length))) * (i_micro_kernel_config->datatype_size_out)
```
please remove comments when no longer needed

deeptiag1 replied:
Removed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -882,7 +891,7 @@ void libxsmm_x86_instruction_vec_move( libxsmm_generated_code* io_generated_code
 
   /* check that we are not masking 'y' */
-  if ( (io_generated_code->arch < LIBXSMM_X86_AVX512) && (i_mask_reg_number != 0) ) {
+  if ( (io_generated_code->arch < LIBXSMM_X86_AVX512) && (io_generated_code->arch != LIBXSMM_X86_AVX512_VL256) && (i_mask_reg_number != 0) ) {
```
Just change LIBXSMM_X86_AVX512 to LIBXSMM_X86_AVX512_VL256

deeptiag1 replied:
Fixed

alheinecke commented on src/generator_x86_instructions.c:
```diff
@@ -909,11 +918,15 @@ void libxsmm_x86_instruction_vec_move( libxsmm_generated_code* io_generated_code
       unsigned int l_encoder_instr = ((i_vmove_instr >> 30) & 0x03);
 
       /* determine encoder */
+      //Change by D-
```
is here a change missing as there is only comment about a change, but not change?

deeptiag1 replied:
I had added an change but reverted it, so this was a bad comment. Removed it

alheinecke commented on src/generator_gemm_sse_avx_avx2_avx512.c:
```diff
@@ -805,6 +844,20 @@ LIBXSMM_API_INTERN unsigned int libxsmm_generator_gemm_sse_avx_avx2_avx512_updat
+      libxsmm_generator_gemm_init_micro_kernel_config_fullvector( io_micro_kernel_config, i_arch, i_xgemm_desc, l_use_masking_a_c );
+    } else {
+      /* we are done with m_blocking */
+    }
```
F64 branch for AVX512_VL256 seems to be missing

sbhatta3 replied:
added

alheinecke commented on src/generator_gemm_sse_avx_avx2_avx512.c:
```diff
@@ -843,7 +896,7 @@ LIBXSMM_API_INTERN unsigned int libxsmm_generator_gemm_sse_avx_avx2_avx512_get_m
                                                                                                const unsigned int                  i_arch ) {
   if ( i_arch <= LIBXSMM_X86_ALLFEAT ) {
-    if ( i_arch >= LIBXSMM_X86_AVX512 ) {
+    if ( i_arch >= LIBXSMM_X86_AVX512_VL256) { // change by D-
```
special handling for int8 and bf16 seems to be missing in the subsequent lines, can be fixed later

deeptiag1 replied:
Int8 and Bf16 still works with the existing if conditions. I am not sure new check is required here.

alheinecke reviewed (changes requested):
I missed changes to following routine:
https://github.com/hfp/libxsmm/blob/master/src/generator_gemm_common.c#L1565
so everything requiring masking won't work.

alheinecke commented on src/generator_gemm_sse_avx_avx2_avx512.c:
```diff
@@ -637,7 +637,7 @@ LIBXSMM_API_INTERN unsigned int libxsmm_generator_gemm_sse_avx_avx2_avx512_get_i
     }
 //Change by D-
-  } else if ( ( i_arch == LIBXSMM_X86_AVX512_VL256 ) && ( ( LIBXSMM_GEMM_PRECISION_F32  == LIBXSMM_GETENUM_OUT( i_xgemm_desc->datatype ) )  ||
+  } else if ( ( i_arch <= LIBXSMM_X86_AVX512_VL256 ) && ( ( LIBXSMM_GEMM_PRECISION_F32  == LIBXSMM_GETENUM_OUT( i_xgemm_desc->datatype ) )  ||
```
<= looks suspicious, should be ==

sbhatta3 replied:
fixed

sbhatta3 replied:
fixed

sbhatta3 replied:
would be good to move it up, right after LIBXSMM_X86_AVX2 check.

alheinecke commented on src/generator_gemm.c:
```diff
@@ -48,12 +48,10 @@ void libxsmm_generator_gemm_kernel( libxsmm_generated_code*        io_generated_
     l_vector_length = 4;
   } else if ( ( io_generated_code->arch <= LIBXSMM_X86_AVX2 ) && LIBXSMM_GEMM_PRECISION_F32 == LIBXSMM_GETENUM_INP( l_xgemm_desc_mod.datatype ) ) {
     l_vector_length = 8;
-  } else if ( ( io_generated_code->arch <= LIBXSMM_X86_AVX512_VL256 ) && LIBXSMM_GEMM_PRECISION_F64 == LIBXSMM_GETENUM_INP( l_xgemm_desc_mod.datatype ) ) {
```
why was this removed? That looked right, perhaps == instead of <=?

sbhatta3 replied:
added back

alheinecke commented on src/generator_gemm.c:
```diff
@@ -48,12 +48,10 @@ void libxsmm_generator_gemm_kernel( libxsmm_generated_code*        io_generated_
-    l_vector_length = 8;
   } else if ( ( io_generated_code->arch <= LIBXSMM_X86_ALLFEAT ) && LIBXSMM_GEMM_PRECISION_F64 == LIBXSMM_GETENUM_INP( l_xgemm_desc_mod.datatype ) ) {
     l_vector_length = 8;
+  } else if ( ( io_generated_code->arch <= LIBXSMM_X86_AVX512_VL256 ) && LIBXSMM_GEMM_PRECISION_F32 == LIBXSMM_GETENUM_INP( l_xgemm_desc_mod.datatype ) ) {
```
check for VL256 should before general AVX512 checks as it was before.

sbhatta3 replied:
fixed

alheinecke reviewed (changes requested):
see further comments

deeptiag1 commented:
Following changes in this  [6679b23] commit:
1. Incorporated @alheinecke  feedback.
2. Removed C++ styled comments.
3. Tested  I16I32, USI8I32, SUI8I32, SUI8UI8 for NN_EQLD sizes.  
4. BF16F32, BF16, BF1632_FLAT, BF16_FLAT works for sizes <= 32x32.

Next Steps: 
1. I16I32, USI8I32, SUI8I32, SUI8UI8  datatypes needs to be tested for following use case:
      - NN_GTLD
      - NT_EQLD
      - NT_GTLD
2.  Enable BF16 for sizes BF16F32, BF16, BF1632_FLAT, BF16_FLAT >32x32 and fix masking issues.

deeptiag1 commented:
Closing this pull request, Thanks for the feedback!
